File: my-folder/0024-swap-nodes-in-pairs/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode cur = head;
        if(cur == null) {
            return cur;
        }
        ListNode swap = head.next;
        if(swap == null) {
            return cur;
        }
        
        cur.next = swapPairs(swap.next);
        swap.next = cur;
        return swap;
        /*ListNode prev = null;
        head = swap;
        
        while(cur.next != null && swap.next != null) {
            cur.next = swap.next;
            swap.next = cur;
            if(prev != null) {
                prev.next = swap;
            }
            prev = cur;
            cur = cur.next;
            swap = cur.next;
        }
        if(swap != null) {
            cur.next = swap.next;
            swap.next = cur;
            if(prev != null) {
                prev.next = swap;
            }
        }
        
        return head;*/
    }
}

File: my-folder/0024-swap-nodes-in-pairs/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs(head: ListNode | null): ListNode | null {
    // base case where less than 2 elements left
    if (!head || !head.next) {
        return head;
    }
    // swap
    const temp = head.next;
    head.next = swapPairs(temp.next)
    temp.next = head;
    return temp;
};
